docs/indexing.md

```diff
@@ -0,0 +1,422 @@
+
+$(d_0, d_1) \mapsto (d_0, d_1)$ for $\boldsymbol{d} \in [0, 2] \times [0, 49]$
+
+- output -> input 2:
+
+$(d_0, d_1) \mapsto (d_0, d_1 - 50)$ for $\boldsymbol{d} \in [0, 2] \times [50, 79]$
+
+The inputs to output map:
+
+- input 1 -> output: $(d_0, d_1) \mapsto (d_0, d_1)$ for $\boldsymbol{d} \in {\rm Dom}(input_1)$.
+- input 2 -> output: $(d_0, d_1) \mapsto (d_0, d_1 + 50)$ for $\boldsymbol{d} \in {\rm Dom}(input_2)$.
+
+### Dot (output-to-input implemented
+
+Indexing maps for dot are very similar to the ones of reduce.
+
+```c+
+p0 = f32[4, 128, 256] parameter(0)
+p1 = f32[4, 256, 64] parameter(1)
+dot = f32[4, 128, 64] dot(p0, p1), 
+  lhs_batch_dims={0}, rhs_batch_dims={0},
+  lhs_contracting_dims={2}, rhs_contracting_dims={1}
+```
+The output to inputs maps:
+
+- output -> input_1: $(d_0, d_1, d_2) \mapsto (d_0, d_1, s_0)$ for $\boldsymbol{d} \in {\rm Dom}(output)$ and $\boldsymbol{s} \in [0, 255]$
+- output -> input_2: $(d_0, d_1, d_2) \mapsto (d_0, s_0, d_2)$ for $\boldsymbol{d} \in {\rm Dom}(output)$ and $\boldsymbol{s} \in [0, 255]$
+
+The inputs to output maps:
+
+- input_1 -> output: $(d_0, d_1, d_2) \mapsto (d_0, d_1, s_0)$ for $\boldsymbol{d} \in {\rm Dom}(input_1)$ and $\boldsymbol{s} \in [0, 63]$
+- input_2 -> output: $(d_0, d_1, d_2) \mapsto (d_0, s_0, d_1)$ for $\boldsymbol{d} \in {\rm Dom}(input_2)$ and $\boldsymbol{s} \in [0, 127]$
+
+### Reduce-window (TBD)
+
+### Pad (TBD)
+
+
+
+## Indexing Maps for Fusion
+
+Indexing map for fusion op is a composition of indexing maps for every op in the cluster. It can happen that some inputs are read several times with different access patterns.
+
+### One input, several indexing maps
+
+Here is an example for $p_0 + p_0^T$
+
+```c+
+f {  
+  p0 = f32[1000, 1000] parameter(0)
+  transpose_p0 = f32[1000, 1000]{0, 1}
+  transpose(p0), dimensions={1, 0}
+  ROOT a0 = f32[1000, 1000] add(p0, transpose_p0)
+}
+```
+
+The output-to-input indexing maps for `p0` will be $(d_0, d_1) \mapsto (d_0, d_1)$ and $(d_0, d_1) \mapsto (d_1, d_0)$.
+
+### One input, deduplicated indexing map
+
+![img](./images/indexing_analysis_transposes.svg)
+
+There are cases when the indexing maps are actually the same, even though it is not immediately obvious. 
+
+```c+
+f {   
+  p0 = f32[20, 10, 50] parameter(0)
+  lhs_transpose_1 = f32[10, 20, 50] transpose(p0), 
+    dimensions={1, 0, 2}
+  lhs_e = f32[10, 20, 50] exponential(lhs_transpose_1)
+  lhs_transpose_2 = f32[10, 50, 20] transpose(lhs_e),
+    dimensions={0, 2, 1}   
+  rhs_transpose_1 = f32[50, 10, 20] transpose(p0),
+    dimensions={2, 1, 0}   
+  rhs_log = f32[50, 10, 20] exponential(rhs_transpose_1)
+  rhs_transpose_2 = f32[10, 50, 20] transpose(rhs_log),
+    dimensions={1, 0, 2}   
+  ROOT add = f32[10, 50, 20] add(lhs_transpose_2, rhs_transpose_2)
+}
+```
+
+The output-to-input indexing map for `p0` in this case is just $(d_0, d_1, d_2) \mapsto (d_2, d_0, d_1)$.
+
+​          
+
+### Softmax
+
+![img](./images/indexing_analysis_softmax.png)
```

Comment:
The HLO fusion seems to contain duplicate code - is this before DCE?

Reply:
i think it is after.